Recv state transition objects

arminsabouri opened the pull request.

spacebear21 committed:
Move shared response processing code to ohttp.rs

The `process_res` methods across send and receive have a lot of shared
code that can be extracted into reusable functions. `process_get_res`
and `process_post_res` essentially do the same thing, and could
conceivably be combined into one function. They were kept separate
because "202 Accepted" is not a valid response for POST requests.

benalleng committed:
Expand mutant coverage for send directory

This adds the tests to cover mutations for the send directory in
its entirety.

benalleng committed:
Merge parallel feature flag local certificate logic in cli ohttp

These ohttp functions had parallel methods when running tests vs running
in production. This made it difficult to read as there were two
functions that appeared to do completely different things when a feature
flag was present.

arminsabouri committed:
Implement `serde::Deserialize` for `core::version`

This commit adds deserialization support for the version enum so it can
be stored in the session event log.

arminsabouri committed:
Introduce Receiver `SessionEvent`

This commit adds session events generated by processing the receiver
state machine. These events wrap v1 receiver types to avoid duplicating
session context learned during the initial typestate transition. To
enable serialization and deserialization of session events, we implement
serde::Serialize and serde::Deserialize on the underlying v1 types. As a
result, we must enable the bitcoin/serde feature under the v1 flag
because `OptionalParameters` includes a bitcoin::FeeRate field.

To enable comparisons between SessionEvents during tests,
all event types must also implement `PartialEq` and `Eq`.

arminsabouri committed:
Add ser/deser tests for `ReceiverSessionEvent`

This commit adds test coverage for the full
serialization/deserialization round trip for each receiver session
event.

arminsabouri committed:
Replace `Persister` trait with `SessionPersister`

This commit updates the `v2::Receiver` state transition functions to
return state transition objects. Each state transition produces
different outcomes based on the current state. When persisted, a state
transition object returns the next valid state.

Changes to `v2::receiver`

Serialization and Deserialization Error Code

Terminal session events now store JSON replies. These events must derive
`serde::Serialize` and `serde::Deserialize` to support serialization and
deserialization.

`v1/mod.rs`

The session history object enables application developers to introspect
a session and retrieve the Payjoin PSBT after inputs have been
contributed. Changes to `v1/mod.rs` allow the session history manager to
expose the PSBT at this specific state. This update addresses a
requirement from the Liana integration, where the contributed PSBT must
be saved to a separate database.

`session.rs`

This file now contains the core logic for the session history manager,
receiver session events, and associated error types.

`receive/v2/mod.rs`

Introduced a sum type over all possible receiver states. This sum type
processes a session event and, if the event is valid for the current
state, progresses to the next state by calling the typestate's `apply`
function. The `apply` function provides a shorthand mechanism to
transition to the next state based on the session event.

Replaced `NewReceiver` with `UninitializedReceiver` and introduced a
`create_session` method, which takes the same parameters as
`NewReceiver::new`.

Updated each state transition method to match on specific errors. The
system classifies errors as either transient (`ImplementationError`) or
fatal (`SessionError`).

Changes to Payjoin-CLI

Database

Introduced a session wrapper struct to persist session events. This
wrapper will support both sender and receiver sessions. As session
updates occur, the system deserializes the session wrapper, appends the
new event log, and reserializes the session. Each session has a unique
`u64` identifier, derived via `serde`. When a session completes (either
successfully or due to a terminal error), the session persister records
a `completed_at` timestamp.

Application

Refactored the receiver session handling to eliminate compile-time state
assumptions. Introduced a unified handler that matches on any receiver
state and drives the session to completion. Each state handler now
exists as an individual method, and each method invokes the next state
handler until the session completes successfully or encounters a
terminal error.

Lastly, replaced `try_contributing_inputs` with `contribute_inputs` for clarity
and consistency.

benalleng committed:
Cleanup find_change_index error comparison

I got a bit confused on the error comparison when working with the
InternalBuilderSender error and the error comparison here. This is
cleaner to just compare to an Err type and not an Option of the internal
error.

benalleng committed:
Use session enpoint url for directory when available

when merging the ohttp-key fetching and validation functions I found
that there was some logic that was skipping the correct directory
selection.

By ensuring that the directory is chosen per session we can ensure that
the correct directory is used no matter the initial config at
intialization.

arminsabouri committed:
squash: ffi changes

This commit monomorphizes each state transition object to be specific to
its corresponding state transition, removing generic parameters to
enable UniFFI exposure. Each state transition object is wrapped in an
`Arc<RwLock<Option<Object>>>` to allow FFI bindings to reference self
safely. UniFFI does not enforce Rust’s strong borrowing guarantees and
operates more predictabily with `&self`.

Additionally, this commit introduces `JsonReceiverSessionPersister`,
enabling applications to persist and load session events using exposed
to_json and from_json methods on `ReceiverSessionEvent`.

spacebear21 committed:
Make UncheckedProposal::from_request `body` &[u8]

Simplify the function signature to take a bytes slice instead of a
generic reader. `parse_body` is renamed to better reflect its role.

spacebear21 committed:
Make V1Context::process_response `response` &[u8]

Simplify the function signature to take a bytes slice instead of a
generic reader.

spacebear21 committed:
Remove InternalValidationError::Io variant

They are no longer constructed after the changes from the previous two
commits.

arminsabouri committed:
squash: Refactor Error Request Extraction and Response Handling

This commit refactors the extraction of error requests and the handling
of responses. Previously, these methods were only available on the
`UncheckedProposal` typestate. However, a session may fail at any
typestate, and the receiver must send an error response to the sender if
the failure is fatal.

The extract_err_req and process_err_res functions are now implemented as
pure functions. By replaying the session event log, applications can
call `extract_err_req` at any point. If the session is in a terminal
state, the function exposes the associated Request and OHTTP context.

spacebear21 committed:
Refactor DirectoryResponseError mappings

Instead of mapping DirectoryResponseError to the corresponding variants
on InternalSessionError and InternalCreateRequestError, nest those
variants by replacing them with DirectoryResponseError.

arminsabouri committed:
Add test coverage for `SessionHistory` and replays

This commit adds test coverage for `replay_receiver_event_log` and the
session history object produced by replaying the logs. The tests persist
session event logs and replay them to verify that the resulting state
matches the expected state.

arminsabouri committed:
Export `InMemoryTestPersister` under `_test-utils`

Co-authored-by: Dan Gould <d@ngould.dev>

`InMemoryTestPerister` now becomes generic over what events it is
storing in memory allowing for other rust-payjoin mods to test storing
and replaying session logs. We expose the in-memory session persister
under a new feature flag: `_test-utils`.

We also re-export the in-memory session persister through the
payjoin-test-utils crate and added a crates-io patch
making it available to other workspace members that use
payjoin-test-utils at version 0.0.0.

0xBEEFCAF3 commented on payjoin-cli/src/db/v2.rs:
```diff
@@ -36,50 +55,91 @@ impl Persister<Sender<WithReplyKey>> for SenderPersister {
-                serde_json::from_slice(&value).map_err(Error::Deserialize)?;
-            sessions.push(session);
+            let (key, _) = item?;
+            session_ids.push(SessionId::new(u64::from_be_bytes(key.as_ref().try_into().unwrap())));
```
Remove unwrap

0xBEEFCAF3 commented on payjoin/src/receive/v1/mod.rs:
```diff
@@ -765,6 +765,9 @@ impl ProvisionalProposal {
     }
+
+    /// Return the PSBT with the receiver's inputs contributed
+    pub fn psbt(&self) -> bitcoin::Psbt { self.payjoin_psbt.clone() }
```
This can be crate

0xBEEFCAF3 commented on payjoin-ffi/src/receive/uni.rs:
```diff
@@ -461,48 +744,60 @@ impl PayjoinProposal {
-pub trait ReceiverPersister: Send + Sync {
-    fn save(&self, receiver: Arc<WithContext>) -> Result<Arc<ReceiverToken>, ForeignError>;
-    fn load(&self, token: Arc<ReceiverToken>) -> Result<Arc<WithContext>, ForeignError>;
+pub trait JsonReceiverPersistedSession: Send + Sync {
```
Should be `SessionPersister`

spacebear21 committed:
Move shared response processing code to ohttp.rs (#660)

The `process_res` methods across send and receive have a lot of shared
code that can be extracted into reusable functions. `process_get_res`
and `process_post_res` essentially do the same thing, and could
conceivably be combined into one function. They were kept separate
because "202 Accepted" is not a valid response for POST requests.

shinghim committed:
Introduce constructors for legacy input pairs

spacebear21 committed:
Merge parallel feature flag local certificate logic in cli ohttp (#726)

These ohttp functions had parallel methods when running tests vs running
in production. This made it difficult to read as there were two
functions that appeared to do completely different things when a feature
flag was present.

Because localhost relays are not actually able to be proxies here we
still need to opportunistically select them from the config when running
our tests instead of just allowing local cert validation.
```
[2025-05-30T17:17:53Z DEBUG payjoin_cli::app::v2::ohttp] Failed to connect to relay: http://localhost:34391/, Internal(InternalError(Reqwest(reqwest::Error { kind: Request, url: Url { scheme: "https", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("payjo.in")), port: None, path: "/.well-known/ohttp-gateway", query: None, fragment: None }, source: hyper_util::client::legacy::Error(Connect, "unsuccessful tunnel") })))
```

spacebear21 committed:
Introduce constructors for legacy input pairs (#753)

Checking off some remaining tasks in #583 - this PR introduces a
constructor for each legacy type

spacebear21 committed:
Avoid unnecessary allocation of request body

There is no real need to copy it just to return it back to the caller.

spacebear21 committed:
Simplify v1 response processing function signatures (#745)

This was prompted by the discussion in
https://github.com/payjoin/rust-payjoin/pull/710#discussion_r2104814421.

Using a reader still left callers susceptible to writing memory
exhaustion vulnerable code, and just introduced extra complexity by
having to buffer twice. We can recognize this and leave the buffering
entirely up to the caller. Also, the V2 sender and receiver response
processing functions already take &[u8], so this makes the function
signatures more consistent across versions.

benalleng committed:
Lock cargo mutants version to 25.0.1

The new version of cargo mutants introduced a set of new mutants that
ultimately is not worth the added effort to fix at the moment as well as
introducing some breaking changes that have not been clarified yet how
to work around them. For now I am locking the cargo mutants version
until we have more clarity on the stability of the mutants workflow and
we can put more priority into full mutation coverage.

spacebear21 committed:
Lock cargo mutants version to 25.0.1 (#759)

The new version of cargo mutants introduced a set of new mutants that
ultimately is not worth the added effort to fix at the moment as well as
introducing some breaking changes that have not been clarified yet how
to work around them. For now I am locking the cargo mutants version
until we have more clarity on the stability of the mutants workflow and
we can put more priority into full mutation coverage.

This should resolve the mutants found in #752 for now. However, as we
move up to future versions we will need to get coverage for these
eventually.

spacebear21 committed:
Receiver Session Events (#760)

This PR adds session events generated by processing the receiver
state machine. These events wrap v1 receiver types to avoid duplicating
session context learned during the initial typestate transition. To
enable serialization and deserialization of session events, we implement
serde::Serialize and serde::Deserialize on the underlying v1 types. As a
result, we must enable the bitcoin/serde feature under the v1 flag
because `OptionalParameters` includes a bitcoin::FeeRate field.

To enable comparisons between SessionEvents during tests,
all event types must also implement `PartialEq` and `Eq`.

Tests cover roundtrip ser/deserialization of each session event.


These commits were cherry-picked off #750 and cleaned up.

benalleng committed:
Add all of the send dir to the cargo workflow

With full mutation coverage we can now include the whole send directory
in our mutants toml.

arminsabouri committed:
Capture hpke reply key in session event

The session reply key stays `None` until we parse it from an
`UncheckedProposal`. We lose this information because we don’t store an
updated `SessionContext` in the `UncheckedProposal` session event and we
don’t want to store the full updated context either. Instead, we stick
to storing only what the session learns at that point. In this case,
that’s the proposal itself and the optional HPKE reply key.

arminsabouri committed:
Export `InMemoryTestPersister` under `_test-utils` (#761)

`InMemoryTestPerister` now becomes generic over what events it is
storing in memory allowing for other rust-payjoin mods to test storing
and replaying session logs. We expose the in-memory session persister
under a new feature flag: `_test-utils`.

We also re-export the in-memory session persister through the
payjoin-test-utils crate and added a crates-io patch making it available
to other workspace members that use payjoin-test-utils at version 0.0.0.
This exposes the in-memory session persister to the int. and FFI test
suites.

arminsabouri committed:
Capture hpke reply key in session event (#762)

The session reply key stays `None` until we parse it from an
`UncheckedProposal`. We lose this information because we don’t store an
updated `SessionContext` in the `UncheckedProposal` session event and we
don’t want to store the full updated context either. Instead, we stick
to storing only what the session learns at that point. In this case,
that’s the proposal itself and the optional HPKE reply key.

This change is cherry-picked off #750

benalleng committed:
Deduplicate multiparty response error with ohttps DirectoryResponse

In #660 we added a process_get_res and process_post_res methods in
ohttps to pull away some of the logic from our send module but
multiparty remained the same.

This PR attempts pull sync the multiparty sender error handling with
this new approach. As well this change gets rid of the remaining mutant
on this status code match block.

spacebear21 committed:
Full mutation coverage on send in the payjoin crate (#748)

This adds full mutation coverage for our sender. After this merges we
will have full mutation coverage on send and receive! A nice step
forward for adversarial tests!

benalleng committed:
Remove FinalizeResponseError in favor of using FinalizedError

After #745 I realized we can go further and completely remove the
FinalizeResponseError in favor of using FinalizedError which already has
access to DirectoryResponse errors.

spacebear21 committed:
Update README test checklist

Having eliminated all mutants in the sender and receiver modules I think
it's safe to claim we have "many unit tests".

spacebear21 committed:
Remove FinalizeResponseError in favor of using FinalizedError (#764)

After #745 I realized we can go further and completely remove the
`FinalizeResponseError` in favor of using `FinalizedError` which already
has access to `DirectoryResponse` errors.

DanGould committed:
Update README test checklist (#765)

benalleng committed:
Combine test_output_substitution with sender checklist test for clarity

By combining the test_output_substitution test and
test_process_proposal_when_payee_output_has_disallowed_output_substitution
we can deduplicate some test logic and ensure that these tests cover all
of the checklist test vectors.

spacebear21 committed:
Combine test_output_substitution with sender checklist test for clarity (#767)

By combining the `test_output_substitution` test and
`test_process_proposal_when_payee_output_has_disallowed_output_substitution`
we can de-duplicate some test logic and ensure that these tests cover
all of the checklist test vectors.

Hopefully we can get all the checks completed in our
[checklist](https://github.com/payjoin/rust-payjoin/issues/339#issuecomment-2895966159)
with these.

benalleng committed:
Correct cargo mutants version locking in weekly workflow

spacebear21 committed:
Correct cargo mutants version locking in weekly workflow (#776)

Corrects the incorrect syntax introduced in #759

arminsabouri committed:
Sender Session Events

This commit adds session events generated by processing the sender
state machine. To enable serialization and deserialization of session events,
we implement `serde::Serialize` and `serde::Deserialize` on the underlying
v1 & v2 types that are captured in session events.

To enable comparisons between `SessionEvents` during tests,
all event types must also implement `PartialEq` and `Eq`.

Tests cover roundtrip ser/deserialization of each session event.

All of our uniffi exports are under the same namespace (see
payjoin_ffi.udl). This commit adds a sender/reciever name prefix to
the exported session event object avoid a naming conflict.

[Related ticket](https://github.com/payjoin/rust-payjoin/issues/723)

arminsabouri committed:
Remove unused imports in python int. tests

arminsabouri committed:
Remove unused imports in python int. tests (#779)

Remove unused imports from the python integration tests.

arminsabouri committed:
Update state transitions to return perisistedError

FFI state transitions objects were returning `ImplementationError` which made no
use of our FFI exported receiver error types. The new `PersistedError`
type will wrap around api side error if they occur and use
`ImplementationError` in the case of an application storage related
error

arminsabouri committed:
Remove storage error wrapper

This wrapper aimed to reduce boilerplate when converting generic
storage errors into `FFI::PersisterError`. However, since this
conversion occurs in only a few places, maintaining the wrapper adds
unnecessary overhead in rust-payjoin.

arminsabouri committed:
Remove storage error wrapper (#781)

This wrapper aimed to reduce boilerplate when converting generic storage
errors into `ffi::{receive|send}::PersisterError`. However, since this
conversion occurs in only a few places, maintaining the wrapper adds
unnecessary overhead in rust-payjoin.

arminsabouri committed:
Sender Session Events (#777)

This PR adds session events generated by processing the sender state
machine. To enable serialization and deserialization of session events,
we implement `serde::Serialize` and `serde::Deserialize` on the
underlying v1 & v2 types that are captured in session events.

To enable comparisons between `SessionEvents` during tests, all event
types must also implement `PartialEq` and `Eq`.

Tests cover roundtrip ser/deserialization of each session event.

arminsabouri committed:
Rename receiever persist.rs to session.rs

Renamed `persist.rs` to `session.rs` as the file no longer manages
persistence logic. It now handles session replays and history
management.
